user_manual.md

```diff
@@ -245,120 +289,157 @@ In particular, this means that any computation that pattern matches `d` will not
+In the above example, `(and x y z)` is treated as `(and (and x y) z)`.
 
-### <a name="assoc-nil"></a>Right/Left associative with nil terminator
+Note that the type for right and left associative operators is typically `(-> T T T)` for some type `T`.
+More generally, a constant declared with the `:right-associative` annotation must have a type of the form `(-> T1 T2 T2)` for some types `T1` and `T2`. Similarly, a constant declared with the `:left-associative` annotation must have a type of the form `(-> T1 T2 T1)`.
+
+<a name="assoc-nil"></a>
+
+#### Right/Left associative with nil terminator
 
 Eunoia supports a variant of the aforementioned functionality where a (ground) nil terminator is provided.
 
-```
+```smt
 (declare-const or (-> Bool Bool Bool) :right-assoc-nil false)
 (define P ((x Bool) (y Bool) (z Bool)) (or x y z))
 (define Q ((x Bool) (y Bool)) (or x y))
 (define R ((x Bool) (y Bool)) (or x))
 ```
 
-In the above example, `(or x y z)` is syntax sugar for `(or x (or y (or z false)))`,
-`(or x y)` is syntax sugar for `(or x (or y false))`,
-and `(or x)` is syntax sugar for `(or x false)`.
+In the above example, `(or x y z)` is treated as `(or x (or y (or z false)))`,
+`(or x y)` is treated as `(or x (or y false))`,
+and `(or x)` is treated as `(or x false)`.
+<!--CT added -->In contrast, if `or` was annotated with `left-associative-nil`,
+`(or x y z)` would be treated as `(or (or (or false x) y) z)`,
+`(or x y)` as `(or (or false x) y)`,
+and `(or x)` as `(or false x)`.
 
-The advantage of right (resp. left) associative operators nil terminators is that the terms they specify are unambiguous, in constrast to operators without nil terminators.
+The advantage of right or left associative operators with nil terminators is that the terms they specify are unambiguous, which is not the case for right or left associative operators without nil terminators.
 In particular, note the following example:
 
-```
+```smt
 (declare-const or (-> Bool Bool Bool) :right-assoc-nil false)
 (define P ((x Bool) (y Bool) (z Bool)) (or x (or y z)))
 (define Q ((x Bool) (y Bool) (z Bool)) (or x y z))
 ```
 
-If `or` would have been marked `:right-assoc`, then the definition of both `P` and `Q` would be `(or x (or y z))` after desugaring.
+<!--CT We should stress that we get to different AST for the P and Q above.
+-->
+
+ <!--CT We should point out that the nil element should be a left and right identity of the assoc symbol, otherwise non-intuitive results occur. 
+
+For instance, with
+(declare-const + (-> Int Int Int) :right-assoc-nil 1)
+We have that 
+(+ x (+ y z))) = (+ x (+ (+ y (+ z 1)) 1)) = x + y + z + 2
+is not equivalent to
+(+ x (+ y (+ z 1))) = x + y + z + 1
+However, I would use this example not to argue that it is advantageous to use the :[left|right]-associative-nil annotation but to warn the reader that the nil element better be a left/right identity for the binary operator, otherwise the semantic is highly unintuitive.
```

Comment:
Ok, I would suggest we add an additional example after this one to demonstrate this.

Reply:
Will do.